### PR TITLE
Node-ctrl Server Basics

### PR DESCRIPTION
**Pull request by AhmedSoliman**
Node-ctrl Server Basics

Introduces a node-level control service listening on port 5121 (to pair with the future node-to-node port will be on 5122)
The node-ctrl service also initializes and exposes prometheus metrics. Components are advised to collect metrics with
the `metrics` crate, and the recorder will automatically inject labels from the current `tracing` span (merged with parent spans).

To avoid an explosion of time series on high-cardinality span attributes, an allow list is defined in `node-ctrl/src/metrics.rs`
that will be expanded over time.

For every crate, we should create `src/metric_definitions.rs` that defines `describe_metrics()`. This function describes the metrics
so that it appears in prometheus metric explorer with help messages. Note that it's not a requirement to describe metrics before using
but it helps to centralize all metrics maintained by every crate in a single place as demonstrated in this PR.

Features:
- Node-ctrl configuration is under `node_ctrl` key in config file. Most notably, `node_ctrl.disable_prometheus` will stop prometheus recording.
- `node_ctrl.histogram_inactivity_timeout` (disabled by default) will evict idle histograms if idle after certain timeout to reclaim memory.
- Prometheus scraping is supported on `GET http://localhost:5122/metrics` (try with curl).
- A stub NodeCtrl GRPC service is put in place multiplexing on the same port based on content-type.

Side Notes:
 - The changes in Cargo.toml files are due to a newer version of test-log that requires every macro user to have `test-log` crate available.
 - The current defined metrics are meant to act as an example and will _almost certainly_ change in the future.

Example:
```
❯ http localhost:5122/metrics
HTTP/1.1 200 OK
content-length: 1167
content-type: text/plain; charset=utf-8
date: Wed, 10 Jan 2024 11:43:13 GMT

# TYPE partition_storage_tx_created_total counter
partition_storage_tx_created_total{partition_id="690"} 1

# TYPE invoker_invocation_task_started_total counter
invoker_invocation_task_started_total{rpc_service="interpreter.CommandInterpreter"} 5
invoker_invocation_task_started_total{rpc_service="verifier.CommandVerifier"} 30

# TYPE invoker_invocation_task_failed_total counter
invoker_invocation_task_failed_total{rpc_service="interpreter.CommandInterpreter",transient="true"} 5
invoker_invocation_task_failed_total{rpc_service="verifier.CommandVerifier",transient="true"} 30

# TYPE ingress_request_created_total counter
ingress_request_created_total{rpc_service="verifier.CommandVerifier",rpc_method="Execute"} 1

# TYPE partition_apply_command_total counter
partition_apply_command_total{partition_id="690",command="ServiceInvocation"} 1

# TYPE partition_storage_tx_committed_total counter
partition_storage_tx_committed_total{partition_id="690"} 1

# TYPE invoker_inflight_invocations_total gauge
invoker_inflight_invocations_total{rpc_service="verifier.CommandVerifier"} 0
invoker_inflight_invocations_total{rpc_service="interpreter.CommandInterpreter"} 0

```

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1050).
* #1063
* #1062
* __->__ #1050

### CODE REVIEW

**Comment by AhmedSoliman** [comment]
@slinkydeveloper suggested to call this the `Node API` and I like that. I will update the PR to use this name. This makes it easy to distinguish between the node admin and cluster admin. Cluster Admin Server will be called "Admin Server".

**Comment by AhmedSoliman** [comment]
> @slinkydeveloper suggested to call this the `Node API` and I like that. I will update the PR to use this name. This makes it easy to distinguish between the node admin and cluster admin. Cluster Admin Server will be called "Admin Server".

We landed on `node-ctrl` as the name of choice. The PR has been updated accordingly.

File: crates/tracing-instrumentation/src/lib.rs [hunk 1]
```diff
@@ -287,7 +288,11 @@ impl Options {
         let layers = tracing_subscriber::registry();
 
         // Logging layer
-        let layers = layers.with(self.log.build_layer()?);
+        let layers = layers
+            .with(self.log.build_layer()?)
+            // Enables auto extraction of selected span labels in emitted metrics.
+            // allowed labels are defined in restate_node_ctrl::metrics::ALLOWED_LABELS.
+            .with(MetricsLayer::new());
```
Comment (tillrohrmann) [inline crates/tracing-instrumentation/src/lib.rs:295]: Really cool :-)

File: crates/worker/src/partition/state_machine/effect_interpreter.rs [hunk 2]
```diff
@@ -167,6 +169,7 @@ where
         let Self { txn, collector } = self;
 
         txn.commit().await?;
+        counter!(PARTITION_STORAGE_TX_COMMITTED).increment(1);
```
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/effect_interpreter.rs:172]: When rebasing on `main`, this increment should no longer be needed.

**Review by tillrohrmann — APPROVED** [review]
Thanks a lot for creating this PR @AhmedSoliman. The changes look really good to me. +1 for merging after resolving the minor comments. Will now try it out to see some metrics reported by Restate :-)

File: Cargo.toml [hunk 3]
```diff
@@ -111,15 +115,17 @@ serde_with = "2.2"
 serde_yaml = "0.9"
 sync_wrapper = "0.1.2"
 tempfile = "3.6.0"
+test-log = { version = "0.2.11", default-features = false, features = ["trace"] }
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply Cargo.toml:119]: The code generated off the macro has changed recently (`cargo update` broke it). It now references `::test_log` so it doesn't work with our re-export anymore.

```
let _ =  ::test_log::tracing_subscriber::FmtSubscriber::builder().with_env_filter(::test_log::tracing_subscriber::EnvFilter::from_default_env()).with_span_events(__internal_event_filter).with_test_writer().try_init();
```

File: crates/worker/src/partition/state_machine/effect_interpreter.rs (hunk 2 above)
Comment (AhmedSoliman) [reply crates/worker/src/partition/state_machine/effect_interpreter.rs:172]: Care to elaborate on why?

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply Cargo.toml:119]: Ah ok. Then we probably can remove the re-export from `test-util` as a follow-up.

[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply Cargo.toml:119]: #1063

**Pull request closed without merging**